src/main/java/org/commonjava/indy/service/archive/controller/ArchiveController.java

```diff
@@ -0,0 +1,319 @@
+            FileInputStream fis = new FileInputStream( artifact );
+            String entryPath = artifact.getPath().split( contentBuildDir )[1];
+
+            zip.putNextEntry( new ZipEntry( entryPath ) );
+
+            int length;
+            while ( ( length = fis.read(buffer ) ) > 0 )
+            {
+                zip.write( buffer, 0, length );
+            }
+            zip.closeEntry();
+            fis.close();
+        }
+        zip.close();
+
+        //clean obsolete build contents
+        for ( File artifact : artifacts )
+        {
+            Files.delete( artifact.toPath() );
+        }
+        dir.delete();
+        return Optional.of( part );
+    }
+
+    public void renderArchive( File part, final String buildConfigId )
+    {
+        if ( !part.exists() )
+        {
+            return;
+        }
+        final File target = new File( archiveDir, buildConfigId + ARCHIVE_SUFFIX );
+        target.delete();
+        target.getParentFile().mkdirs();
+        part.renameTo( target );
+    }
+
+    public Optional<File> getArchiveInputStream( final String buildConfigId ) throws IOException
+    {
+        File targetDir = new File( archiveDir );
+        if ( !targetDir.exists() )
+        {
+            return Optional.empty();
+        }
+
+        List<File> contents = walkedAllFiles( archiveDir );
+        for ( File content : contents )
+        {
+            if ( content.getName().equals( buildConfigId + ARCHIVE_SUFFIX ) )
+            {
+                return Optional.of( content );
+            }
+        }
+        return Optional.empty();
+    }
+
+    public void deleteArchive( final String buildConfigId ) throws IOException
+    {
+        File targetDir = new File( archiveDir );
+        if ( targetDir.exists() )
+        {
+            List<File> contents = walkedAllFiles( archiveDir );
+            for ( File content : contents )
+            {
+                if ( content.getName().equals( buildConfigId + ARCHIVE_SUFFIX ) )
+                {
+                    content.delete();
+                }
+            }
+            return;
+        }
+    }
+
+    private List<File> walkedAllFiles ( String path ) throws IOException
+    {
+        List<File> contents = Files.walk( Paths.get( path ) )
+                .filter( Files::isRegularFile )
+                .map( Path::toFile )
+                .collect( Collectors.toList() );
+        return contents;
+    }
+
+    private void fileTrackedContent( String contentBuildDir, final HistoricalContentDTO content )
+    {
+        File tracked = new File( contentBuildDir, content.getBuildConfigId() );
+        tracked.getParentFile().mkdirs();
+        try
+        {
+            String json = objectMapper.writeValueAsString( content );
+            FileOutputStream out = new FileOutputStream( tracked );
+            IOUtils.copy( new ByteArrayInputStream( json.getBytes() ), out );
```

Comment:
Seems these two streams are not closed, use try(resource) here?

Reply:
I think is been taken care by destroy() above

Reply:
@shokakucarrier These are totally different stuff. The destroy() is closing the httpclient, but these two streams are not related.

Reply:
@ligangty to amend

Reply:
@yma88 Just saw the commit. One last thing is this ByteArrayInputStream. If you checked the IOUTils.copy, you will see there is no close action for both in and out. So that means this one will also need to be closed  manually.